[JSC] `toIntegerOrInfinity` should trunc a negative fractional to `+0.0`

sosukesuzuki opened the pull request:
#### 1d296b968d89cb37715dd09be81e31f270afeb28
<pre>
[JSC] `toIntegerOrInfinity` should trunc a negative fractional to `+0.0`
<a href="https://bugs.webkit.org/show_bug.cgi?id=294734">https://bugs.webkit.org/show_bug.cgi?id=294734</a>

Reviewed by Ross Kirsling.

According to the specification, `ToIntegerOrInfinity` should truncate negative
fractional values to `+0.0`[1]. However, the current JSC implementation truncates
them to `-0.0`, which may cause errors in some operations.

Therefore, this patch changes `ToIntegerOrInfinity` to truncate negative
fractional values to `+0.0`.

[1]: <a href="https://tc39.es/ecma262/#sec-tointegerorinfinity">https://tc39.es/ecma262/#sec-tointegerorinfinity</a>

* JSTests/stress/to-integer-or-infinity-tranc-negative-fractional.js: Added.
(shouldBe):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
(JSC::JSValue::toIntegerOrInfinity const):

Canonical link: <a href="https://commits.webkit.org/296468@main">https://commits.webkit.org/296468@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9bce38a18a49a7d530378683b992c2f3b076674

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108574 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28235 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18659 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113783 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58970 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28924 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36789 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82471 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111522 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22951 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97796 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62907 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22373 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15935 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58497 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/101131 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92321 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15986 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116903 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/107133 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35628 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26266 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91496 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36001 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94067 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91298 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23268 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36185 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13950 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31453 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35530 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41065 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/131415 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35242 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35651 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38590 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36919 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
```diff
@@ -133,7 +133,9 @@ inline double JSValue::toIntegerOrInfinity(JSGlobalObject* globalObject) const
-    return trunc(std::isnan(d) ? 0.0 : d + 0.0);
+    if (std::isnan(d))
+        return 0.0;
+    return trunc(d) + 0.0;
```
`trunc(d + 0.0)` for consistency?

rkirsling commented on JSTests/stress/to-integer-or-infinity-tranc-negative-fractional.js:
```diff
@@ -0,0 +1,10 @@
+function shouldBe(a, b) {
```
nit: filename has `tranc` instead of `trunc`

sosukesuzuki replied:
Changed to use `isnan(d) ? 0.0 : trunc(d) + 0.0` everywhere for consistency.

sosukesuzuki committed:
[JSC] `toIntegerOrInfinity` should trunc a negative fractional to `+0.0`
https://bugs.webkit.org/show_bug.cgi?id=294734

Reviewed by Ross Kirsling.

According to the specification, `ToIntegerOrInfinity` should truncate negative
fractional values to `+0.0`[1]. However, the current JSC implementation truncates
them to `-0.0`, which may cause errors in some operations.

Therefore, this patch changes `ToIntegerOrInfinity` to truncate negative
fractional values to `+0.0`.

[1]: https://tc39.es/ecma262/#sec-tointegerorinfinity

* JSTests/stress/to-integer-or-infinity-tranc-negative-fractional.js: Added.
(shouldBe):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
(JSC::JSValue::toIntegerOrInfinity const):

Canonical link: https://commits.webkit.org/296468@main
